bb.edn

```diff
@@ -0,0 +1,82 @@
+{
+ :paths ["src"]
+ :deps {zprint/zprint  {:mvn/version "1.2.9"}}
+ :tasks
+ {
+  :requires
+  [[build.build :as build]
+   [build.recurse-keys :as recurse]
+   [babashka.fs :as bfs]
+   [godot-java.core :as godot-java]
+   [babashka.nrepl.server :as srv]
+   ]
+
+  :init
+  (do
+    (def opts {:pre-start-fn build/echo})
+    (def godot-java-jar "target/godot_java-1.0-SNAPSHOT-jar-with-dependencies.jar")
+    (def vars ;; being data allows other tasks to manipulate this map more easily
+      {:JAVA_HOME (build/java-home)
+       :INCLUDES ["-I":JAVA_HOME"/include -I":JAVA_HOME"/include/":OS]
+       :LIBS ["-ljvm -L ":JAVA_HOME"/lib -L ":JAVA_HOME"/lib/server"]
+
+       :compile ["gcc -g -Wall -fPIC ":INCLUDES" -c src/entry.c -o entry.o"]
+       :link ["gcc -g -fPIC ":INCLUDES" ":LIBS" -shared -o entry.so entry.o -rdynamic"]})
+    )
+
+  make-macos
+  {:doc "Equivalent of make for macos"
+   :task (let [{:keys [compile link]} (-> vars (merge {:OS "darwin"})
+                                          (recurse/recurse-keys) #_ (doto build/zpprint))]
```

Comment:
Take a look here re `recurse/recurse-keys`.
It offers a data orientated way to keep on replacing keywords until only strings are left.

I was surprised I couldn't find something that does this already, but took 5 minutes with ChatGPT (and a small correction because it got it got the code wrong). At some point I want this in a library for myself, but I did it first here so hey ho. Of course if you know of a tool that does this then please let me know and I'll replace.

IMO this is more flexible and concise than bindings like `(str "gcc -g -fPIC" INCLUDES "...")`. Unlike doing this with variable bindings, you can offer a substitution later as with `:OS`, which actually makes it super flexible.

Reply:
Oh so it's like a more advanced string formatting. My only problem about recurse-keys is that it took me quite a bit of effort to understand what is going on and I feel like this solution is more clever than it should be.

I checked out babashka libs and found out babashka comes with selmer (selmer.parser, https://github.com/yogthos/Selmer ). Since it's mostly about placing values inside strings, I think it will work well. This quote from readme summarizes the project well:

> Selmer templates consist of plain text that contains embedded expression and filter tags. While Selmer is primarily meant for HTML generation, it can be used for templating any text.